buildroot-external/patches/occu/0158-WebUI-Fix-SystemprotocolCharacterEncoding.patch

```diff
@@ -0,0 +1,10 @@
+--- occu/WebUI/www/rega/esp/functions.fn.orig
++++ occu/WebUI/www/rega/esp/functions.fn
+@@ -58,6 +58,7 @@
+           sRet = sRet.Replace("\r\n", "<br/>");
+           sRet = sRet.Replace("\r", "<br/>");
+           sRet = sRet.Replace("\n", "<br/>");
++          sRet = sRet.Replace("'","&#39;");
```

Comment:
Wenn wir schon dabei sind, sollten wir auch mal andere Sonderzeichen da testen und auch so sachen wie ein tabulator (0x9) usw. und dann entsprechend auch escapen. Und natürlich sollte man ggf. auch rausfinden warum ein `'` überhaupt an der Stelle ein Problem ist. Das kommt mir immer noch komisch vor das man das escapen muss. Und was ist mit dieser `::WriteDPText()` Funktion? Wird die wirklich nur an dieser Stelle verwendet oder ist das eine allgemeine die auch von anderen stellen verwendet wird. Weil dann könnte das ggf. zu problemen führen.

Reply:
>  warum ein ' überhaupt an der Stelle ein Problem ist. Das kommt mir immer noch komisch vor das man das escapen muss.

Weil die Rückgabe dann wieder in einen String gepackt wird, der mit `'` umschlossen ist.
```
Write("tmp['value'] = '"#sCollectedValues#"';");
```

z.B.
```
Write("tmp['value'] = 'TEXT'TEXT1';");
```

https://github.com/eq-3/occu/blob/68c6c5e57d13ca2cbdb3d8f6c5d0ee7edb41ef4d/WebUI/www/rega/pages/tabs/control/systemProtocolLoader.htm#L49

Reply:
Verstehe. Aber dann wäre es doch eigentlich "richtiger" statt ein `'` meit einem `&#39;` zu ersetzen, stattdessen ein `\'` draus zu machen. Und wie gesagt fehlt auch noch in der Ersetzungsliste der `\t` (Tabulator) oder was passiert wenn der da ist?

Reply:
Gerade mal getestet: Wenn ein Tab im Text ist, wird das Systemprotokoll auch nicht korrekt dargestellt.

Reply:
Ich habe eine Editor genommen, einen Text geschrieben mit Tabzeichen, kopiert und in die SV eingefügt.
Beim Editieren des SV Inhalts in der WebUI ist das Tab-Zeichen auch nach dem Speichern weiterhin vorhanden.

Lässt man sich das Log jedoch im Skripteditor ausgeben, wird am Tab-Zeichen umgebrochen.
Der Inhalt der Variable selbst wird jedoch korrekt dargestellt

Scheinbar hat `GetHistoryData` ein Problem mit dem Tab ?

![grafik](https://user-images.githubusercontent.com/18190139/145340005-8f04493b-b780-42f2-b7bc-74c22ab1a55a.png)

Reply:
Oder liegt es am `foreach`?

Reply:
wieso soll das am `foreach` liegen?

Reply:
Achso, wegen dem `\t` als trenner? das kann natürlich gut sein, ja.

Reply:
> Achso, wegen dem \t als trenner? 

Ja und wie ich der eQ3-Skript-Doku entnehme, ist `foreach` auf `\t` festgenagelt, oder kann man als Parameter ein "Wunschtrennzeichen" mitgeben?

Reply:
Nein, das kann man nicht. Deshalb muss man bevor man dort daten reinfüttert die selber `\t` beinhalten könnten diese in was anderes wandeln und danach ggf. zurück.

Reply:
Hmm, das müsste dann ja innerhalb der ReGa passieren, denn man hat von außen ja keinen Einfluss auf den String von "protokollierten" SVen, wenn sie in die interne History wandern.

Reply:
Naja, es ist ja schon richtig das die ReGa den Inhalt der SV 1:1 in die History übernimmt. Das Problem ist eher die Kombination aus `foreach()` und `GetHistoryData()` und das dabei übersehen wurde das das GetHistoryData() eben einen string mit tabs nicht nur zwischen den Datentrennern zurückliefern kann, sondern eben auch im Text selbst.

Reply:
Achja, und wenn ich mir das ganze da so anschaue wie das GetHistoryData() verwendet wird, dann kommt der sicherlich auch schnell durcheinander wenn der Text der Systemvariablen ein `;` Zeichen enthält weil die Routine zum auseinandernehmen das `;` als Spaltentrenner einsetzt.

Reply:
Also eigentlich müsste man das so in der Art machen:

```c++
string s;
integer rCount;
integer iStart=0;
integer iCount=dom.GetHistoryDataCount();
string history=dom.GetHistoryData( iStart, iCount, &rCount );
history = history.Replace("\t", "XXX");
history = history.Replace(";XXX", "\t");
foreach(s, history)
{
  WriteLine(s.Replace("XXX", "\t"));
}
```

Allerdings stellen sich da meine Nackenhaare ein wenig auf und mir fallen dann gleich (eben auch das `;`) noch andere Character ein die ggf. ein Problem sein könnten. Und was ist wenn im Text jetzt das "XXX" zu finden ist? usw. usw. Sicherlich lösbar, nur nicht 100% IMHO.

Reply:
Korrekt. Sieht dann so aus wie mit \t. Aber auch beim ; fällt mir spontan keine Lösung ein.

Reply:
Vmtl. wäre es wirklich das beste das direkt in der ReGa im `GetHistoryData()` anzugehen und dort vor ausgabe das `\t` und `;` sicherheitshalber in was anderes (was genau?) zu wandeln sodass GetHistoryData() schon gar nicht ein `\t` oder `;` ausliefert.

Reply:
> sicherheitshalber in was anderes (was genau?) zu wandeln

Vielleicht in `[ph_tab]` und `[ph_semi]` _(ph = placeholder) als Abkürzung für mehr Eindeutigkeit_ und wir ersetzen die Zeichenfolge anschließend vor der finalen (Web-)ausgabe (wie bereits`'` hier im Patch in der `functions.fn`)?

Reply:
Ne, ich bin jetzt soweit das ich das bereits in ReGa umgesetzt habe durch eine einfache suche&ersetze funktion die alle semikolon und tabs entsprechend ersetzen (aus `;` wird `,` und aus `\t` wird ` `). Weil die Funktion wird ja auch z.B. von externen Anwendungen/Apps usw. verwendet und da passiert dann das gleiche wie bei der WebUI. Und den Tot muss man dann halt sterben das der text selbst kein ; oder \t beinhalten kann.

Reply:
So, mit der kommenden ReGaHss `R1.00.0388.0228` sollte es nun entsprechend behoben sein. Danke nochmal für das Abstubsen :)